crates/red_knot_python_semantic/src/types.rs

```diff
@@ -521,6 +521,53 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Resolves the boolean value of a type.
+    /// This is used to determine the value that would be returned when `__bool__` is called on an object
+    pub fn bool(&self, db: &'db dyn Db) -> Option<bool> {
+        match self {
+            Type::Any => None,
+            Type::Never => None,
+            Type::Unknown => None,
+            Type::Unbound => None,
+            Type::None => Some(false),
+            Type::Function(_) | Type::RevealTypeFunction(_) => Some(true),
+            Type::Module(_) => Some(true),
+            Type::Class(_) => Some(true),
+            Type::Instance(_) => None,
+            Type::Union(union) => {
+                let inner_values = union
+                    .elements(db)
+                    .iter()
+                    .map(|elem| elem.bool(db))
+                    .collect::<FxOrderSet<_>>();
+                let mut found_true = false;
+                let mut found_false = false;
+                for value in inner_values {
+                    match value {
+                        Some(true) => found_true = true,
+                        Some(false) => found_false = true,
+                        None => return None,
+                    }
+                }
```

Comment:
I think you should be able to iterate over the union elements directly here without collecting, which should be cheaper:

```suggestion
                let mut found_true = false;
                let mut found_false = false;
                for element in union.elements(db) {
                    match element.bool(db) {
                        Some(true) => found_true = true,
                        Some(false) => found_false = true,
                        None => return None,
                    }
                }
```